pkg/cloudprovider/openstack.go

```diff
@@ -0,0 +1,848 @@
+			return false, err
+		}
+
+		for _, p := range portList {
+			if has, err := hasNodeName(p.Description, nodeName); err != nil || !has {
+				continue
+			}
+			for _, fip := range p.FixedIPs {
+				if fip.SubnetID == s.ID && fip.IPAddress == ip.String() {
+					ports = append(ports, p)
+					// End the search.
+					return false, nil
+				}
+			}
+		}
+		// Get the next list of ports from the pager.
+		return true, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(ports) != 1 {
+		return nil, fmt.Errorf("Expected to find a single port, instead found %d ports", len(ports))
+	}
+	return &ports[0], nil
+}
+
+// allowIPAddressOnNeutronPort adds the specified IP address to the port's allowed_address_pairs.
+func (o *OpenStack) allowIPAddressOnNeutronPort(portID string, ip net.IP) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Always get the most recent copy of this port.
+		p, err := neutronports.Get(o.neutronClient, portID).Extract()
+		if err != nil {
+			return err
+		}
+
+		// Sanity check to see if the IP is already inside the port's allowed_address_pairs.
+		if isIPAddressAllowedOnNeutronPort(*p, ip) {
+			return AlreadyExistingIPError
+		}
+
+		// Update the port's allowed_address_pairs by appending to it.
+		// According to the neutron API:
+		// "While the ip_address is required, the mac_address will be taken from the port if not specified."
+		// https://docs.openstack.org/api-ref/network/v2/index.html?expanded=update-port-detail
+		allowedPairs := append(p.AllowedAddressPairs, neutronports.AddressPair{
+			IPAddress: ip.String(),
+		})
+		// Update the port. Provide the revision number to make use of neutron's If-Match
+		// header. If the port has received another update since we last retrieved it, the
+		// revision number won't match and neutron will return a "RevisionNumberConstraintFailed"
+		// error message.
+		opts := neutronports.UpdateOpts{
+			AllowedAddressPairs: &allowedPairs,
+			RevisionNumber:      &p.RevisionNumber,
+		}
+		_, err = ports.Update(o.neutronClient, p.ID, opts).Extract()
+
+		// If the update yielded an error of type "RevisionNumberConstraintFailed", then create a
+		// Conflict error. RetryOnConflict will react to this and will repeat the entire operation.
+		if err != nil && strings.Contains(err.Error(), "RevisionNumberConstraintFailed") {
+			return &apierrors.StatusError{
+				ErrStatus: metav1.Status{
+					Message: err.Error(),
+					Reason:  metav1.StatusReasonConflict,
+					Code:    http.StatusConflict,
+				},
+			}
+		}
+
+		// Any other error or nil, return.
+		return err
+	})
+}
+
+// unallowIPAddressOnNeutronPort removes the specified IP address from the port's allowed_address_pairs.
+func (o *OpenStack) unallowIPAddressOnNeutronPort(portID string, ip net.IP) error {
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		// Always get the most recent copy of this port.
+		p, err := neutronports.Get(o.neutronClient, portID).Extract()
```

Comment:
Question: is openstack somehow deletes the port, should we be able to handle that gracefully here?
Not sure if this is too much of a pathological case, but thought of asking so this can be as robust as possible.